src/main/java/org/apache/maven/shared/incremental/IncrementalBuildHelper.java

```diff
@@ -358,6 +368,79 @@ public void afterRebuildExecution( IncrementalBuildHelperRequest incrementalBuil
 
     }
 
+    private void writeChangedFiles(
+            DirectoryScanner directoryScanner,
+            String[] filesBeforeAction,
+            File mojoConfigBase,
+            String createdFilesListFileName ) throws MojoExecutionException
+    {
+        if ( directoryScanner.getBasedir() == null )
+        {
+            return;
+        }
+
+        DirectoryScanResult outputDirectoryScanResult = scanDirectoryDiff(
+                directoryScanner, filesBeforeAction );
+
+        try
+        {
+            writeChangedFiles( outputDirectoryScanResult, mojoConfigBase, createdFilesListFileName );
+        }
+        catch ( IOException e )
+        {
+            throw new MojoExecutionException( "Error while storing the mojo status", e );
+        }
+    }
+
+    private static void writeChangedFiles(
+            DirectoryScanResult outputDirectoryScanResult,
+            File mojoConfigBase,
+            String createdFilesListFileName )
+            throws IOException
+    {
+        File createdFiles = new File( mojoConfigBase, createdFilesListFileName );
+        String[] filesAdded = outputDirectoryScanResult.getFilesAdded();
+        String filesAddedAsString = Stream.of( filesAdded ).collect( Collectors.joining( "\n" ) );
+
+        Files.write(
+                createdFiles.toPath(),
+                filesAddedAsString.getBytes( StandardCharsets.UTF_8 ),
+                StandardOpenOption.CREATE );
+    }
+
+    private static DirectoryScanResult scanDirectoryDiff(
+            DirectoryScanner directoryScanner,
+            String[] filesBeforeAction )
+    {
+        // now scan the same directory again and create a diff
+        directoryScanner.scan();
+        DirectoryScanResult outputScanResult =
+                directoryScanner.diffIncludedFiles( filesBeforeAction );
+        return outputScanResult;
+    }
+
+    private static List<String> deleteFiles( File fileNameIndex, File parent ) throws IOException
+    {
+        List<String> oldFiles = Files.readAllLines( fileNameIndex.toPath() );
+        for ( String oldFileName : oldFiles )
+        {
+            File oldFile = new File( parent, oldFileName );
+            oldFile.delete();
+        }
+        return oldFiles;
+    }
+
+    private String[] scanDirectory( DirectoryScanner directoryScanner, File directory )
+    {
+        directoryScanner.setBasedir( directory );
```

Comment:
This is a little worrisome. I didn't realize directoryScanner was mutable. This might suggest not using a directoryScanner field at all and just creating one when we need it.